tachyon/math/base/byinverter.h

```diff
@@ -0,0 +1,469 @@
+#ifndef TACHYON_MATH_BASE_BYINVERTER_H_
+#define TACHYON_MATH_BASE_BYINVERTER_H_
+
+namespace tachyon::math {
+/// Signed 64-bit integer type
+typedef __int64_t i64;
+/// Unsigned 64-bit integer type
+typedef __uint64_t u64;
+/// Unsigned 128-bit integer type
+typedef __uint128_t u128;
+/// Unsigned integer type for storing object sizes
+typedef size_t usize;
+
+/// Returns the non-largest argument
+constexpr usize least(usize a, usize b) { return a < b ? a : b; }
+
+/// Returns the number of trailing zeros in "n"
+/// represented in the little-endian binary code
+constexpr usize trailing0(u64 n) { return n != 0 ? __builtin_ctzll(n) : 64; }
+
+/// Big signed (B * L)-bit integer type, whose variables
+/// store numbers in the two's complement code as arrays
+/// of B-bit chunks, where B < 64 and L > 0. The ordering
+/// of the chunks in these arrays is little-endian. The
+/// arithmetic operations for this type are wrapping ones
+template <usize B, usize L> struct CInt {
+ public:
+  /// Mask, in which the B lowest bits are 1 and only they
+  constexpr static u64 MASK = ~0UL >> (64 - B);
+
+  /// Returns the number equal to the argument. The binary
+  /// representation of the absolute value of the argument
+  /// must have no more than B * L - 1 bits to ensure that
+  /// the result and the argument represent the same number
+  constexpr static CInt load(const i64 input) {
+    CInt result;
+    u64 raw = input;
+    CInt::convert<64, B>(&raw, 1, result.data, L);
+    if ((B * L > 64) && (input < 0)) {
+      // For the two's complement code and n > m the n-bit representation
+      // of a negative number can be expressed as its m-bit representation
+      // preceded by m - n bits equal to 1
+      result.data[64 / B] |= (~0UL << (64 % B)) & CInt::MASK;
+      for (usize i = 64 / B + 1; i < L; i++) {
+        result.data[i] = CInt::MASK;
+      }
+    }
+    return result;
+  }
+
+  /// Returns the number equal to the one specified by the arguments. Its
+  /// absolute value is (64 * "isize")-bit integer stored as the array of
+  /// 64-bit chunks, whose address is specified by "input". The ordering
+  /// of the chunks in the array is little-endian. A non-zero input number
+  /// is negative iff "sign" is "true". The binary representation of the
+  /// absolute value of the input number must have no more than B * L - 1
+  /// bits to ensure that the result and arguments specify the same number
+  constexpr static CInt load(const u64 *input, const usize isize, bool sign) {
```

Comment:
Please use `BigInt<N>` instead of `input` and `isize`, also include the boundary check at compile time.

```
  template <typename N>
  constexpr static CInt load(const BigInt<N>& input, bool sign) {
    static_assert(N * 64 <= B * L);
  }
```

Reply:
Your BigInt is unsigned, isn't it?

Reply:
`  template <size_t N>
  constexpr static CInt Load(const BigInt<N> &input, bool sign) {
    CInt result;
    Convert<64, B>(input.limbs, N, result.data, L);
    return sign ? -result : result;
  }`

For this data type I would like to allow loading from the larger BigInts: some using parties may know that certain BigInt variable may contain some small value due to features of an algorithm.